### ISSUE TITLE
识别模型 训练和推理效果一致，实际部署识别效果不一致。

### ISSUE THREAD

**Issue by kingkingpang** [PaddlePaddle/PaddleOCR#6736]
请提供下述完整信息以便快速定位问题/Please provide the following information to quickly locate the problem

- 系统环境/System Environment：Windows10_Qt5.12_MSVC_C++ ——光功率计数码管字符识别项目
- 版本号/Version：python 3.7 Paddle：PaddlePaddle 2.3.0  PaddleOCR：PP-OCRv3 问题相关组件/Related components：rec
-重新修改了字典如下：
<img width="902" alt="截屏2022-06-28 下午8 23 23" src="https://user-images.githubusercontent.com/53334346/176177793-e4521c04-d977-447e-99f7-135430c5c479.png">

数据量不大，大概150张，在预训练模型上微调。
因此增加了训练轮次，增大了学习率。yaml如下：

Global:
  debug: false
  use_gpu: false
  epoch_num: 200
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/v3_en_mobile
  save_epoch_step: 3
  eval_batch_step: [0, 50]
  cal_metric_during_train: true
  pretrained_model:
  checkpoints:
  save_inference_dir:
  use_visualdl: false
  infer_img: doc/imgs_words/ch/word_1.jpg
  character_dict_path: ppocr/utils/dict/digit_dict.txt
  max_text_length: &max_text_length 25
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ./output/rec/predicts_ppocrv3_en.txt

Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.01
    warmup_epoch: 5
  regularizer:
    name: L2
    factor: 3.0e-05

Architecture:
  model_type: rec
  algorithm: SVTR
  Transform:
  Backbone:
    name: MobileNetV1Enhance
    scale: 0.5
    last_conv_stride: [1, 2]
    last_pool_type: avg
  Head:
    name: MultiHead
    head_list:
      - CTCHead:
          Neck:
            name: svtr
            dims: 64
            depth: 2
            hidden_dims: 120
            use_guide: True
          Head:
            fc_decay: 0.00001
      - SARHead:
          enc_dim: 512
          max_text_length: *max_text_length

Loss:
  name: MultiLoss
  loss_config_list:
    - CTCLoss:
    - SARLoss:

PostProcess:  
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc
  ignore_space: False

Train:
  dataset:
    name: SimpleDataSet
    data_dir: ./train/
    ext_op_transform_idx: 1
    label_file_list: ["./train/rec_gt.txt"]
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - RecConAug:
        prob: 0.5
        ext_data_num: 2
        image_shape: [48, 320, 3]
    - RecAug:
    - MultiLabelEncode:
    - RecResizeImg:
        image_shape: [3, 48, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_sar
        - length
        - valid_ratio
  loader:
    shuffle: true
    batch_size_per_card: 32
    drop_last: true
    num_workers: 4
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ./train/
    label_file_list: ["./train/test.txt"]
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - MultiLabelEncode:
    - RecResizeImg:
        image_shape: [3, 48, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_sar
        - length
        - valid_ratio
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 32
    num_workers: 4

训练过程和推理过程根据文档来都很流畅，效果也不错。
识别图片：
![11](https://user-images.githubusercontent.com/53334346/176178806-876bdd8b-4bae-414c-afc5-42500719d77d.jpg)

训练后预测结果：
[2022/06/28 13:45:56] ppocr INFO: train with paddle 2.3.0 and device Place(cpu)
[2022/06/28 13:45:57] ppocr INFO: load pretrain successful from ./output/v3_en_mobile/best_accuracy
[2022/06/28 13:45:57] ppocr INFO: infer_img: ./doc/imgs_words/ch/11.jpg
[2022/06/28 13:45:57] ppocr INFO: 	 result: 1.7047	0.8836452960968018
[2022/06/28 13:45:57] ppocr INFO: success!

推理后预测结果：
[2022/06/28 14:37:50] ppocr INFO: In PP-OCRv3, rec_image_shape parameter defaults to '3, 48, 320', if you are using recognition model with PP-OCRv2 or an older version, please set --rec_image_shape='3,32,320
[2022/06/28 14:37:50] ppocr INFO: Predicts of ./doc/imgs_words/ch/11.jpg:('1.7047', 0.8836452960968018)

- 实际部署后的结果：
![43F08236AAF33E5A13A48EBF6760294C](https://user-images.githubusercontent.com/53334346/176179561-b4454888-e66c-4494-9639-40d60050f8d2.png)
此处 0.01 貌似是输出结果（不确定）。完全不一致。

我在实际部署中只用了识别模型，因此与检测模型无关。

检查了输入图片用opencv读取的，也是BGR&HWC格式，resize为[3,48,320],与配置文件是一致的。
因此个人认为与size应该也没什么关系，实际读入图片尺寸基本也都是31*131。

所以我在想是不是字典的问题？？实际部署字典如下：
![497BD409D394246898043DAF6655610A](https://user-images.githubusercontent.com/53334346/176180947-5692a4d8-b228-4538-b35e-933a01f977d5.png)

相关部分代码如下：
cv::Mat img = cv::imread("./11.jpg");
    // 创建默认配置对象
    paddle_infer::Config config;
    
    // 设置预测模型路径，即为本小节第2步中下载的模型
    //    std::string prog_file = "ch_PP-OCRv3_rec_infer/inference.pdmodel";
    //    std::string params_file = "ch_PP-OCRv3_rec_infer/inference.pdiparams";
    std::string prog_file = "627infer/inference.pdmodel";
    std::string params_file = "627infer/inference.pdiparams";
    config.SetModel(prog_file, params_file);
    auto predictor = paddle_infer::CreatePredictor(config);
    
    // 获取输入 Tensor
    auto input_names = predictor->GetInputNames();
    auto input_tensor = predictor->GetInputHandle(input_names[0]);
    
    // 设置输入 Tensor 的维度信息
    std::vector<int> INPUT_SHAPE = {1, 3, 48, 320};
    input_tensor->Reshape(INPUT_SHAPE);
    
    // 准备输入数据*img->bits()
    int input_size = 1 * 3 * 48 * 320;
    std::vector<float> input_data(input_size, *img.data);
    // 设置输入 Tensor 数据
    input_tensor->CopyFromCpu(input_data.data());
    
    
    predictor->Run();
    
    auto output_names = predictor->GetOutputNames();
    auto output_tensor = predictor->GetOutputHandle(output_names[0]);
    
    // 获取 Output Tensor 的维度信息
    std::vector<int> output_shape = output_tensor->shape();
    int output_size = std::accumulate(output_shape.begin(), output_shape.end(), 1,
                                      std::multiplies<int>());
    
    // 获取 Output Tensor 的数据
    std::vector<float> output_data;
    output_data.resize(output_size);
    output_tensor->CopyToCpu(output_data.data());
    
    auto predict_shape = output_shape;
    qDebug()<<output_shape;
    qDebug()<<output_data;
    //    ui->label_2->setText();
    
    // ctc decode
    std::vector<std::string> str_res;
    int argmax_idx;
    int last_index = 0;
    float score = 0.f;
    int count = 0;
    float max_value = 0.0f;
    
    for (int n = 0; n < predict_shape[1]; n++) {
        argmax_idx =
                int(MainWindow::argmax(&output_data[n * predict_shape[2]],
                &output_data[(n + 1) * predict_shape[2]]));
        max_value =
                float(*std::max_element(&output_data[n * predict_shape[2]],
                &output_data[(n + 1) * predict_shape[2]]));
        
        if (argmax_idx > 0 && (!(n > 0 && argmax_idx == last_index))) {
            score += max_value;
            count += 1;
            str_res.push_back(label_list_[argmax_idx]);
        }
        last_index = argmax_idx;
    }
    score /= count;
    std::string str;
    for (int i = 0; i < str_res.size(); i++) {
        std::cout << str_res[i];
        str += str_res[i];
    }
    std::cout << "\tscore: " << score << std::endl;

**Comment by kingkingpang** [comment]
感觉像是图像预处理的问题，在C++示例中，还有resize（按原图比例缩放）、normalize（归一化？）、permute（转置？）三个操作。

this->resize_op_.Run(srcimg, resize_img, wh_ratio, this->use_tensorrt_);

this->normalize_op_.Run(&resize_img, this->mean_, this->scale_,
                          this->is_scale_);

std::vector<float> input(1 * 3 * resize_img.rows * resize_img.cols, 0.0f);

this->permute_op_.Run(&resize_img, input.data());

于是在代码中加入如下代码：
![C82790D9356E45694BC2970187204773](https://user-images.githubusercontent.com/53334346/176866488-51c890b4-2046-49ac-acf7-90c81d690761.png)

结果如下：
![23A5D2C8408F9035C376183E3B7D207D](https://user-images.githubusercontent.com/53334346/176866567-3a5e4256-0ab7-41f2-8ee7-5e529c5d8347.png)

结果貌似是101，还是不对。

**Comment by kingkingpang** [comment]
更新一下进度，重新调整了一下图像预处理部分的 resize
把resize中的 imgH=32 调整为  imgH=48，
把imgW= int(32*wh_ratio) 调整为 imgW= int(48*wh_ratio)
与训练时的尺寸对齐。

发现识别结果有很大改善，在之前的图片上识别结果为：
1.70474  score 0.855862 

这里万万没想到，resize尺寸对结果会有这么大的影响。
测试其他图片后，基本识别率都在80%以上了，

但面对甚至是训练集中的图片，依然大部分会有漏掉识别、或识别错误的情况。

想达到90%以上识别率，结下来该往哪些地方进行优化呢？

**Comment by Topdu** [comment]
1、如果训练集和测试集的图片都是固定尺寸（31*131），并不一定非要使用48*320的size，可以修改训练时的input_size为（32*136）以适应具体场景，同时可以将resize的方式改成非padding的形式：
```
      - RecResizeImg:
          image_shape: [3, 32, 256]
          padding: False  # 默认是True
```
同时推理时，采用固定尺寸的输入，PP-OCRv3默认是动态尺寸的输入。
2、如果训练集很少，可以不是使用蒸馏训练
3、尝试使用灰度图的输入进行训练

**Comment by D715-ky** [comment]
> 更新一下进度，重新调整了一下图像预处理部分的 resize 把resize中的 imgH=32 调整为 imgH=48， 把imgW= int(32_wh_ratio) 调整为 imgW= int(48_wh_ratio) 与训练时的尺寸对齐。
> 
> 发现识别结果有很大改善，在之前的图片上识别结果为： 1.70474 score 0.855862
> 
> 这里万万没想到，resize尺寸对结果会有这么大的影响。 测试其他图片后，基本识别率都在80%以上了，
> 
> 但面对甚至是训练集中的图片，依然大部分会有漏掉识别、或识别错误的情况。
> 
> 想达到90%以上识别率，结下来该往哪些地方进行优化呢？

你好，我现在遇到跟你一样的问题了，在直接使用模型推理时很准确，但是导出模型之后再推理就识别不准、不全，识别率很低，我也是只使用识别模型，然后在初始化PaddleOCR时，写入了与训练时一样的图像大小：rec_image_shape='3,48,320'，还是不行，请问是在哪里修改识别模型的输入图像大小呢？就是在哪里调整 resize、 imgH和imgW呢？非常感谢

**Comment by kingkingpang** [comment]
你在部署推理的时候应该分为三个部分：前处理，推理，后处理，我指的resize指的是前处理中的，图像预处理部分的 resize 。
c++的话，你得自己写这几个操作，resize（按原图比例缩放）、normalize（归一化）
![image](https://github.com/PaddlePaddle/PaddleOCR/assets/53334346/221493ce-e62b-4b69-a9dc-5074e08a2040)
